protocol/state/snapshot.go

```diff
@@ -1,53 +1,110 @@
-// Package state provides types for encapsulating blockchain state.
 package state
 
 import (
+	"fmt"
+
+	"chain/errors"
 	"chain/protocol/bc"
 	"chain/protocol/patricia"
 )
 
 // Snapshot encompasses a snapshot of entire blockchain state. It
-// consists of a patricia state tree and the issuances memory.
+// consists of a patricia state tree and the nonce set.
 //
-// Issuances maps an "issuance hash" to the time (in Unix millis)
-// at which it should expire from the issuance memory.
+// Nonces maps a nonce entry's ID to the time (in Unix millis)
+// at which it should expire from the nonce set.
 type Snapshot struct {
-	Tree      *patricia.Tree
-	Issuances map[bc.Hash]uint64
+	Tree   *patricia.Tree
+	Nonces map[bc.Hash]uint64
+}
+
+func (s *Snapshot) AddNonce(id bc.Hash, expiryMS uint64) error {
+	if s.Nonces[id] >= expiryMS {
+		return fmt.Errorf("conflicting nonce %x", id[:])
+	}
+	s.Nonces[id] = expiryMS
+	return nil
 }
 
-// PruneIssuances modifies a Snapshot, removing all issuance hashes
+func (s *Snapshot) DeleteSpentOutput(id bc.Hash) error {
+	if !s.Tree.Contains(id[:]) {
+		return fmt.Errorf("invalid prevout %x", id[:])
+	}
+	s.Tree.Delete(id[:])
+	return nil
+}
+
+func (s *Snapshot) AddOutput(id bc.Hash) error {
+	return s.Tree.Insert(id[:])
+}
+
+// PruneNonces modifies a Snapshot, removing all nonce IDs
 // with expiration times earlier than the provided timestamp.
-func (s *Snapshot) PruneIssuances(timestampMS uint64) {
-	for hash, expiryMS := range s.Issuances {
+func (s *Snapshot) PruneNonces(timestampMS uint64) {
+	for hash, expiryMS := range s.Nonces {
 		if timestampMS > expiryMS {
-			delete(s.Issuances, hash)
+			delete(s.Nonces, hash)
 		}
 	}
 }
 
 // Copy makes a copy of provided snapshot. Copying a snapshot is an
-// O(n) operation where n is the number of issuance hashes in the
-// snapshot's issuance memory.
-func Copy(original *Snapshot) *Snapshot {
+// O(n) operation where n is the number of nonces in the snapshot's
+// nonce set.
+func (s *Snapshot) Copy() *Snapshot {
```

Comment:
Reverted. An artifact of how this PR evolved (and of other changes that were already reverted).